# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
from pandas.api.types import is_numeric_dtype
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv("step3_output.csv")
del df["Unnamed: 0"]
df

Name       state  summit_elev  \
0                     Hilltop Ski Area      Alaska         2090   
1                  Sunrise Park Resort     Arizona        11100   
2        Yosemite Ski & Snowboard Area  California         7800   
3                     Donner Ski Ranch  California         8012   
4                        June Mountain  California        10090   
..                                 ...         ...          ...   
162                      Hogadon Basin     Wyoming         8000   
163          Sleeping Giant Ski Resort     Wyoming         7428   
164                   Snow King Resort     Wyoming         7808   
165  Snowy Range Ski & Recreation Area     Wyoming         9663   
166                White Pine Ski Area     Wyoming         9500   

     vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0              294      0        0.0          0          0     0       1  ...   
1             1800      0        0.0          0          1     2       3  ...   
2              600      0        0.0          0          0     0       1  ...   
3              750      0        0.0          0          0     0       1  ...   
4             2590      0        0.0          0          2     0       0  ...   
..             ...    ...        ...        ...        ...   ...     ...  ...   
162            640      0        0.0          0          0     0       0  ...   
163            810      0        0.0          0          0     0       1  ...   
164           1571      0        0.0          0          0     1       1  ...   
165            990      0        0.0          0          0     0       1  ...   
166           1100      0        0.0          0          0     0       2  ...   

     SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0                 30.0       30.000000        150.000000       36.0   
1                800.0       80.000000        115.000000       49.0   
2                 88.0      174.873239        110.000000       84.0   
3                505.0       60.000000        163.000000       82.0   
4               1500.0      174.873239        115.103943       58.0   
..                 ...             ...               ...        ...   
162               92.0       32.000000        121.000000       61.0   
163              184.0       18.000000         61.000000       81.0   
164              400.0      250.000000        121.000000       80.0   
165               75.0       30.000000        131.000000       59.0   
166              370.0      174.873239        115.103943       81.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0               69.0     30.000000      34.00000         152.000000   
1              250.0     74.000000      78.00000         104.000000   
2              300.0     47.000000      47.00000         107.000000   
3              400.0     75.000000      75.00000         170.000000   
4              250.0     57.916957      64.16681         128.000000   
..               ...           ...           ...                ...   
162             80.0     48.000000      48.00000          95.000000   
163            310.0     42.000000      42.00000          77.000000   
164            300.0     59.000000      59.00000         123.000000   
165            250.0     49.000000      49.00000         120.053004   
166            150.0     57.916957      49.00000         120.053004   

     NightSkiing_ac  clusters  
0              30.0         2  
1              80.0         1  
2               0.0         1  
3               0.0         1  
4               0.0         1  
..              ...       ...  
162             0.0         1  
163             0.0         1  
164           110.0         1  
165             0.0         1  
166             0.0         1  

[167 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df = pd.concat([df.drop("state", axis=1), pd.get_dummies(df['state'])], axis=1)
df

Name  summit_elev  vertical_drop  trams  \
0                     Hilltop Ski Area         2090            294      0   
1                  Sunrise Park Resort        11100           1800      0   
2        Yosemite Ski & Snowboard Area         7800            600      0   
3                     Donner Ski Ranch         8012            750      0   
4                        June Mountain        10090           2590      0   
..                                 ...          ...            ...    ...   
162                      Hogadon Basin         8000            640      0   
163          Sleeping Giant Ski Resort         7428            810      0   
164                   Snow King Resort         7808           1571      0   
165  Snowy Range Ski & Recreation Area         9663            990      0   
166                White Pine Ski Area         9500           1100      0   

     fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0.0          0          0     0       1       0  ...             0   
1          0.0          0          1     2       3       1  ...             0   
2          0.0          0          0     0       1       3  ...             0   
3          0.0          0          0     0       1       5  ...             0   
4          0.0          0          2     0       0       4  ...             0   
..         ...        ...        ...   ...     ...     ...  ...           ...   
162        0.0          0          0     0       0       1  ...             0   
163        0.0          0          0     0       1       1  ...             0   
164        0.0          0          0     1       1       1  ...             0   
165        0.0          0          0     0       1       3  ...             0   
166        0.0          0          0     0       2       0  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
162             0          0     0        0         0           0   
163             0          0     0        0         0           0   
164             0          0     0        0         0           0   
165             0          0     0        0         0           0   
166             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
162              0          0        1  
163              0          0        1  
164              0          0        1  
165              0          0        1  
166              0          0        1  

[167 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,"AdultWeekend"]

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [5]:
X

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0           2090            294      0        0.0          0          0     0   
1          11100           1800      0        0.0          0          1     2   
2           7800            600      0        0.0          0          0     0   
3           8012            750      0        0.0          0          0     0   
4          10090           2590      0        0.0          0          2     0   
..           ...            ...    ...        ...        ...        ...   ...   
162         8000            640      0        0.0          0          0     0   
163         7428            810      0        0.0          0          0     0   
164         7808           1571      0        0.0          0          0     1   
165         9663            990      0        0.0          0          0     0   
166         9500           1100      0        0.0          0          0     0   

     triple  double  surface  ...  Rhode Island  South Dakota  Tennessee  \
0         1       0        2  ...             0             0          0   
1         3       1        0  ...             0             0          0   
2         1       3        1  ...             0             0          0   
3         1       5        2  ...             0             0          0   
4         0       4        1  ...             0             0          0   
..      ...     ...      ...  ...           ...           ...        ...   
162       0       1        1  ...             0             0          0   
163       1       1        1  ...             0             0          0   
164       1       1        0  ...             0             0          0   
165       1       3        1  ...             0             0          0   
166       2       0        0  ...             0             0          0   

     Utah  Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0       0        0         0           0              0          0        0  
1       0        0         0           0              0          0        0  
2       0        0         0           0              0          0        0  
3       0        0         0           0              0          0        0  
4       0        0         0           0              0          0        0  
..    ...      ...       ...         ...            ...        ...      ...  
162     0        0         0           0              0          0        1  
163     0        0         0           0              0          0        1  
164     0        0         0           0              0          0        1  
165     0        0         0           0              0          0        1  
166     0        0         0           0              0          0        1  

[167 rows x 58 columns]

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [7]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [8]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
y_pred[19]

10746698626253.248

## Review Model Outcomes — Iterate over additional models as needed

In [9]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [10]:
from sklearn.metrics import explained_variance_score,mean_absolute_error
variance = explained_variance_score(y_test, y_pred)
variance

-4.47615775639109e+22

In [11]:
mae = mean_absolute_error(y_test, y_pred)
mae

767616380109.3429

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
lm.intercept_

193053221399.82394

In [13]:
lm.score(X_test, y_test)

-4.820477583774455e+22

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values("Coefficient", ascending = False)[0:10]

Coefficient
fastEight      1.575594e+13
trams          1.081282e+13
fastSixes      9.314082e+12
total_chairs   3.745094e+12
New York       3.412450e+12
Michigan       3.072741e+12
New Hampshire  2.665737e+12
Wisconsin      2.549814e+12
Pennsylvania   2.549814e+12
California     2.295103e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
df2 = pd.read_csv("step3_output.csv")
del df2["Unnamed: 0"]
y = df2["AdultWeekend"]
X = df2.drop(["state", "Name", "AdultWeekend"], axis=1)
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [16]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [17]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [18]:
y_pred = model.predict(X_test)
y_pred

array([52.92036263, 53.38655896, 46.33145573, 65.30910129, 59.96907916,
       64.91869626, 50.11448074, 39.62316253, 45.04712441, 64.65374026,
       53.2496589 , 41.8253599 , 35.354525  , 46.91776828, 53.56306558,
       58.28112709, 45.42601142, 72.10165889, 76.25803934, 56.38778189,
       58.81736666, 49.65666953, 75.55397078, 60.08698981, 52.15262404,
       64.03885442, 56.55359435, 51.56705938, 67.72815744, 60.3745952 ,
       41.59872585, 60.25685472, 59.73313258, 65.4853125 , 62.3037607 ,
       31.26860078, 71.30602562, 34.1092326 , 57.76183058, 77.54153166,
       63.2157856 , 49.75396527])

In [19]:
variance = explained_variance_score(y_test, y_pred)
variance

0.5523999446003682

In [20]:
mae = mean_absolute_error(y_test, y_pred)
mae

6.898348813267889

In [21]:
lm.score(X_test, y_test)

0.5492171017140969

In [22]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values("Coefficient", ascending = False)[0:10]

Coefficient
AdultWeekday        10.919670
summit_elev          3.372629
clusters             2.992936
daysOpenLastYear     1.897315
averageSnowfall      1.751967
Runs                 1.718986
triple               1.345386
vertical_drop        1.280535
surface              1.109235
NightSkiing_ac       0.889761

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [23]:
df3 = pd.read_csv("step3_output.csv")
del df3["Unnamed: 0"]
y = df3["AdultWeekend"]
X = df3.drop(["state", "Name", "AdultWeekend", "summit_elev", "vertical_drop"], axis=1)
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [24]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [25]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [26]:
y_pred = model.predict(X_test)
y_pred

array([54.22845836, 52.58328213, 46.70321625, 65.32157374, 53.97372023,
       64.78425892, 48.73870896, 39.50381093, 44.51107602, 67.00627031,
       52.95745504, 41.16370042, 35.56753191, 47.38751692, 53.82245789,
       58.1503547 , 46.87316694, 68.21676433, 81.06473768, 56.60143111,
       57.82093118, 50.05902697, 76.2865446 , 58.8581852 , 53.93217702,
       63.46449532, 54.94258084, 56.25242595, 67.17777564, 60.11492568,
       41.00585146, 58.85030287, 57.2986834 , 63.59524733, 61.22232274,
       31.23050934, 72.07236223, 32.80208228, 56.103431  , 77.84595359,
       62.99950931, 49.30492276])

In [27]:
variance = explained_variance_score(y_test, y_pred)
variance

0.5516082088038478

In [28]:
mae = mean_absolute_error(y_test, y_pred)
mae

7.100396035723669

In [29]:
lm.intercept_

56.16032724499326

In [30]:
lm.score(X_test, y_test)

0.5460122665753846

In [31]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values("Coefficient", ascending = False)[0:10]

Coefficient
AdultWeekday         10.833653
clusters              2.795036
averageSnowfall       2.708687
Runs                  1.827287
triple                1.330452
daysOpenLastYear      1.324166
NightSkiing_ac        1.270287
SkiableTerrain_ac     1.101911
surface               0.942149
double                0.673144

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -4.47615775639109e+22 | 767616380109.3429 |-|
| Model 2. | 0.5523999446003682| 6.898348813267889|'state'|
| Model 3. | 0.5516082088038478 | 7.100396035723669|'state','summit_elev','base_elev'|

Model Selection:

Model 2 would be my choice as it has the best tradeoff of explained variance and mean absolute error.